near-sdk/src/types/public_key.rs

```diff
@@ -0,0 +1,282 @@
+            Ok(CurveType::SECP256K1)
+        } else {
+            Err(ParsePublicKeyError { kind: ParsePublicKeyErrorKind::UnknownCurve })
+        }
+    }
+}
+
+/// Public key in a binary format with base58 string serialization with human-readable curve.
+/// The key types currently supported are `secp256k1` and `ed25519`.
+///
+/// Ed25519 public keys accepted are 32 bytes and secp256k1 keys are the uncompressed 64 format.
+///
+/// # Example
+/// ```
+/// use near_sdk::PublicKey;
+///
+/// // Compressed ed25519 key
+/// let ed: PublicKey = "ed25519:6E8sCci9badyRkXb3JoRpBj5p8C6Tw41ELDZoiihKEtp".parse()
+///             .unwrap();
+///
+/// // Uncompressed secp256k1 key
+/// let secp256k1: PublicKey = "secp256k1:qMoRgcoXai4mBPsdbHi1wfyxF9TdbPCF4qSDQTRP3TfescSRoUdSx6nmeQoN3aiwGzwMyGXAb1gUjBTv5AY8DXj"
+///             .parse()
+///             .unwrap();
+/// ```
+#[derive(Debug, Clone, PartialEq, PartialOrd, Ord, Eq, BorshDeserialize, BorshSerialize)]
+pub struct PublicKey {
+    data: Vec<u8>,
+}
+
+impl PublicKey {
+    fn split_key_type_data(value: &str) -> Result<(CurveType, &str), ParsePublicKeyError> {
+        if let Some(idx) = value.find(':') {
+            let (prefix, key_data) = value.split_at(idx);
+            Ok((prefix.parse::<CurveType>()?, &key_data[1..]))
+        } else {
+            // If there is no Default is ED25519.
+            Ok((CurveType::ED25519, value))
+        }
+    }
+
+    fn from_parts(curve: CurveType, data: Vec<u8>) -> Result<Self, ParsePublicKeyError> {
+        let expected_length = match curve {
+            CurveType::ED25519 => 32,
+            CurveType::SECP256K1 => 64,
+        };
+        if data.len() != expected_length {
+            return Err(ParsePublicKeyError {
+                kind: ParsePublicKeyErrorKind::InvalidLength(data.len()),
+            });
+        }
+        let mut bytes = Vec::with_capacity(1 + expected_length);
+        bytes.push(curve as u8);
+        bytes.extend(data);
+
+        Ok(Self { data: bytes })
+    }
+
+    /// Returns a byte slice of this `PublicKey`'s contents.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.data
+    }
+
+    /// Converts a `PublicKey` into a byte vector.
+    pub fn into_bytes(self) -> Vec<u8> {
+        self.data
+    }
+
+    /// Get info about the CurveType for this public key
+    pub fn curve_type(&self) -> CurveType {
+        CurveType::from_u8(self.data[0]).unwrap_or_else(|_| unreachable!())
+    }
+}
+
+impl From<PublicKey> for Vec<u8> {
+    fn from(v: PublicKey) -> Vec<u8> {
+        v.data
+    }
+}
+
+impl TryFrom<Vec<u8>> for PublicKey {
+    type Error = ParsePublicKeyError;
+
+    fn try_from(mut data: Vec<u8>) -> Result<Self, Self::Error> {
+        if data.len() == 0 {
+            return Err(ParsePublicKeyError {
+                kind: ParsePublicKeyErrorKind::InvalidLength(data.len()),
+            });
+        }
+
+        let curve = CurveType::from_u8(data.remove(0))?;
+        Self::from_parts(curve, data)
```

Comment:
This is removing the first byte to create the curve type, then adding it back (essentially copying over the data twice), can we not just validate the curve type and initialize it here now that we are just storing the vector?

Reply:
Good catch. I thought I forgot about something when I was transitioning back to an internal `Vec<u8>`